File: reference_python_project/core/dump_hedge_detector.py
"""
╔══════════════════════════════════════════════════════════════════════════════╗
║  DUMP-HEDGE DETECTOR — STRUCTURAL ARBITRAGE ENGINE                           ║
║                                                                              ║
║  Strategy: UP + DOWN tokens always resolve to $1.00 combined.               ║
║  When the market temporarily prices them below $1.00 total, buying both     ║
║  sides locks in a guaranteed profit regardless of direction.                 ║
║                                                                              ║
║  Entry condition: up_price + down_price <= SUM_TARGET (default 0.95)        ║
║  Profit per share pair: 1.00 - combined_price (locked at entry)             ║
║                                                                              ║
║  No Binance data needed. No directional prediction needed.                  ║
║  Pure structural edge from liquidity imbalances in the order book.          ║
╚══════════════════════════════════════════════════════════════════════════════╝
"""

import time
from dataclasses import dataclass
from typing import Dict, List, Optional

from core.market_utils import get_seconds_remaining
from core.polymarket_client import MarketInfo, PolymarketClient
from utils.logger import get_logger

logger = get_logger(__name__)


@dataclass
class DumpHedgeSignal:
    """
    A confirmed dump-hedge arbitrage opportunity.

    Both YES and NO token IDs are included — caller buys BOTH simultaneously.
    Profit is locked at entry: (1.0 - combined_price) * size_shares USDC.
    """
    market: MarketInfo
    asset: str
    yes_token_id: str
    no_token_id: str
    yes_price: float            # current BUY ask for YES (UP) token
    no_price: float             # current BUY ask for NO (DOWN) token
    combined_price: float       # yes_price + no_price
    discount: float             # 1.0 - combined_price (locked profit per share)
    discount_pct: float         # discount / combined_price (ROI at entry)
    seconds_remaining: float    # seconds left in the market window
    timestamp: float

    def __str__(self) -> str:
        return (
            f"DumpHedge | {self.asset.upper()} | "
            f"YES: {self.yes_price:.3f}  NO: {self.no_price:.3f}  "
            f"Sum: {self.combined_price:.3f} | "
            f"Locked: {self.discount:.3f}/share ({self.discount_pct:.1%} ROI) | "
            f"{self.seconds_remaining:.0f}s left"
        )


class DumpHedgeDetector:
    """
    Scans active Polymarket markets for combined YES+NO prices below
    the sum target, indicating a structural arbitrage opportunity.

    Checks all configured assets once per evaluate() call. Returns
    the signal with the highest locked profit, or None.

    Does NOT require Binance data. Works entirely from Polymarket pricing.
    """

    def __init__(
        self,
        polymarket_client: PolymarketClient,
        assets: List[str],
        sum_target: float = 0.95,
        min_discount: float = 0.03,
        min_market_liquidity: float = 1_000.0,
        trade_window_minutes: int = 5,
        min_seconds_remaining: float = 60.0,
        cooldown_seconds: float = 30.0,
    ) -> None:
        self.polymarket_client = polymarket_client
        self.assets = assets
        self.sum_target = sum_target
        self.min_discount = min_discount
        self.min_market_liquidity = min_market_liquidity
        self.trade_window_minutes = trade_window_minutes
        self.min_seconds_remaining = min_seconds_remaining
        self.cooldown_seconds = cooldown_seconds

        self._window_seconds = float(trade_window_minutes * 60)
        self._last_signal_time: Dict[str, float] = {}
        self._signals_generated: int = 0
        self._evaluations: int = 0
        # Mirror of EdgeDetector._active_market_by_asset for dashboard display
        self._active_market_by_asset: Dict[str, object] = {a: None for a in assets}

        logger.info(
            "DumpHedgeDetector initialized | Assets: %s | SumTarget: %.2f | "
            "MinDiscount: %.2f | Window: %dm",
            ", ".join(a.upper() for a in assets),
            sum_target, min_discount, trade_window_minutes,
        )

    async def evaluate(self) -> Optional[DumpHedgeSignal]:
        """
        Scan all configured assets for a dump-hedge opportunity.
        Returns the signal with highest locked discount, or None.
        """
        self._evaluations += 1
        now = time.time()
        best: Optional[DumpHedgeSignal] = None

        for asset in self.assets:
            if now - self._last_signal_time.get(asset, 0.0) < self.cooldown_seconds:
                continue
            signal = await self._evaluate_asset(asset)
            if signal is None:
                continue
            if best is None or signal.discount > best.discount:
                best = signal

        if best:
            self._last_signal_time[best.asset] = time.time()
            self._signals_generated += 1
            logger.info("DUMP-HEDGE DETECTED [#%d]: %s", self._signals_generated, best)

        return best

    async def _evaluate_asset(self, asset: str) -> Optional[DumpHedgeSignal]:
        """
        Check all active markets for an asset and return the one with the
        highest locked discount, or None if no opportunity exists.

        Loops all returned markets instead of only the first so that the bot
        captures the best structural edge when multiple windows are active
        simultaneously (e.g., during a 5-min/15-min transition period).
        """
        markets = await self.polymarket_client.get_active_markets(
            asset=asset,
            min_liquidity=self.min_market_liquidity,
            force_refresh=False,
        )
        if not markets:
            self._active_market_by_asset[asset] = None
            return None

        best: Optional[DumpHedgeSignal] = None

        for market in markets:
            signal = await self._check_market_dh(asset, market)
            if signal is None:
                continue
            if best is None or signal.discount > best.discount:
                best = signal

        # Update dashboard cache with the best market found (or None)
        self._active_market_by_asset[asset] = best.market if best else None
        return best

    async def _check_market_dh(
        self, asset: str, market: object
    ) -> Optional[DumpHedgeSignal]:
        """Evaluate a single market for a dump-hedge opportunity."""
        # Timing check — don't enter too close to expiry
        secs = get_seconds_remaining(market, self._window_seconds)
        if secs is None or secs < self.min_seconds_remaining:
            logger.debug(
                "[%s] DH skip %s: %.0fs remaining < %.0fs minimum",
                asset.upper(), market.question[:40], secs or 0, self.min_seconds_remaining,
            )
            return None

        # Fetch fresh BUY prices for both legs via REST (or WS cache if available).
        # Never fall back to stale discovery-cache prices — they can be minutes old
        # and produce false signals (signal says 0.91, execution sees real 0.99).
        yes_price = await self.polymarket_client.get_market_price(market.yes_token_id, "BUY")
        no_price  = await self.polymarket_client.get_market_price(market.no_token_id,  "BUY")

        if yes_price is None or no_price is None or yes_price <= 0 or no_price <= 0:
            logger.debug(
                "[%s] DH skip %s: price unavailable (yes=%s no=%s)",
                asset.upper(), market.question[:40], yes_price, no_price,
            )
            return None

        combined = yes_price + no_price
        discount = 1.0 - combined

        if combined > self.sum_target:
            logger.debug(
                "[%s] DH: combined=%.4f > target=%.4f — no opportunity",
                asset.upper(), combined, self.sum_target,
            )
            return None

        if discount < self.min_discount:
            logger.debug(
                "[%s] DH: discount=%.4f < min=%.4f — too thin",
                asset.upper(), discount, self.min_discount,
            )
            return None

        return DumpHedgeSignal(
            market=market,
            asset=asset,
            yes_token_id=market.yes_token_id,
            no_token_id=market.no_token_id,
            yes_price=yes_price,
            no_price=no_price,
            combined_price=combined,
            discount=discount,
            discount_pct=discount / combined if combined > 0 else 0.0,
            seconds_remaining=secs,
            timestamp=time.time(),
        )

    def reset_cooldown(self, asset: str) -> None:
        """Reset per-asset cooldown (call after a DH position closes)."""
        self._last_signal_time[asset] = time.time()

    def get_stats(self) -> dict:
        return {
            "dh_evaluations": self._evaluations,
            "dh_signals_generated": self._signals_generated,
        }

File: trading-core/src/signals/DumpHedgeDetector.cpp
#include "DumpHedgeDetector.h"
#include <spdlog/spdlog.h>
#include <fmt/core.h>
#include <algorithm>
#include <cmath>

namespace trading {

namespace {

constexpr double kDebugLogIntervalSec = 30.0;

struct DhNearMiss {
    bool valid = false;
    std::string asset;
    int window_minutes = 0;
    double yes_price = 0.0;
    double no_price = 0.0;
    double combined = 0.0;
    double entry_fees = 0.0;
    double discount = 0.0;
    double seconds_remaining = 0.0;
    std::string reason;
    double score = -1e9;
};

std::string reject_label(const std::string& code) {
    if (code == "window_off") return "window_off";
    if (code == "asset_off") return "asset_off";
    if (code == "cooldown") return "cooldown";
    if (code == "min_time") return "min_time";
    if (code == "no_price") return "no_price";
    if (code == "bad_side") return "bad_side";
    if (code == "sum_high") return "sum_high";
    if (code == "disc_low") return "disc_low";
    return code;
}

void consider_near_miss(DhNearMiss& best, DhNearMiss candidate) {
    if (!candidate.valid) return;
    if (!best.valid || candidate.score > best.score) {
        best = std::move(candidate);
    }
}

} // namespace

static std::string dh_market_key(const MarketInfo& market) {
    return market.asset + "-" + std::to_string(market.window_minutes) + "m";
}

DumpHedgeDetector::DumpHedgeDetector(StateStore& state_store,
                                     std::vector<MarketInfo> active_markets,
                                     double sum_target,
                                     double min_discount,
                                     double min_seconds_remaining,
                                     double cooldown_seconds)
    : state_store_(state_store),
      active_markets_(std::move(active_markets)),
      sum_target_(sum_target),
      min_discount_(min_discount),
      min_seconds_remaining_(min_seconds_remaining),
      cooldown_seconds_(cooldown_seconds)
{
    std::string msg = fmt::format("DumpHedgeDetector initialized | Markets: {} | SumTarget: {:.2f} | MinDiscount: {:.2f}",
                                  active_markets_.size(), sum_target_, min_discount_);
    spdlog::log(spdlog::level::info, msg);
}

std::optional<DumpHedgeSignal> DumpHedgeDetector::evaluate(double current_time_ms) {
    evaluations_++;
    std::optional<DumpHedgeSignal> best_signal;
    DhNearMiss near_miss;

    for (const auto& market : active_markets_) {
        const std::string market_key = dh_market_key(market);
        if (last_signal_time_.contains(market_key)) {
            if ((current_time_ms - last_signal_time_.at(market_key)) / 1000.0 < cooldown_seconds_) {
                continue;
            }
        }

        if (market.window_minutes == 5 && !state_store_.dh_enable_5m()) {
            consider_near_miss(near_miss, DhNearMiss{
                .valid = true, .asset = market.asset, .window_minutes = market.window_minutes,
                .reason = "window_off", .score = -1000.0});
            continue;
        }
        if (market.window_minutes == 15 && !state_store_.dh_enable_15m()) {
            consider_near_miss(near_miss, DhNearMiss{
                .valid = true, .asset = market.asset, .window_minutes = market.window_minutes,
                .reason = "window_off", .score = -1000.0});
            continue;
        }
        if (!state_store_.dh_asset_enabled(market.window_minutes, market.asset)) {
            consider_near_miss(near_miss, DhNearMiss{
                .valid = true, .asset = market.asset, .window_minutes = market.window_minutes,
                .reason = "asset_off", .score = -900.0});
            continue;
        }

        double seconds_remaining = market.end_date_ts - (current_time_ms / 1000.0);
        if (seconds_remaining < min_seconds_remaining_) {
            consider_near_miss(near_miss, DhNearMiss{
                .valid = true, .asset = market.asset, .window_minutes = market.window_minutes,
                .seconds_remaining = seconds_remaining, .reason = "min_time", .score = -800.0});
            continue;
        }

        auto yes_price_opt = state_store_.get_token_price(market.yes_token_id);
        auto no_price_opt = state_store_.get_token_price(market.no_token_id);

        if (!yes_price_opt || !no_price_opt) {
            consider_near_miss(near_miss, DhNearMiss{
                .valid = true, .asset = market.asset, .window_minutes = market.window_minutes,
                .seconds_remaining = seconds_remaining, .reason = "no_price", .score = -700.0});
            continue;
        }

        if (yes_price_opt->side != "BUY" || no_price_opt->side != "BUY") {
            consider_near_miss(near_miss, DhNearMiss{
                .valid = true, .asset = market.asset, .window_minutes = market.window_minutes,
                .yes_price = yes_price_opt->price, .no_price = no_price_opt->price,
                .seconds_remaining = seconds_remaining, .reason = "bad_side", .score = -600.0});
            continue;
        }

        double yes_price = yes_price_opt->price;
        double no_price = no_price_opt->price;

        if (yes_price <= 0 || no_price <= 0) continue;

        double combined = yes_price + no_price;
        double entry_fees = state_store_.compute_dh_entry_fee_per_share(
            yes_price, no_price, market.yes_token_id, market.no_token_id);
        double discount = 1.0 - combined - entry_fees;

        DhNearMiss priced{
            .valid = true,
            .asset = market.asset,
            .window_minutes = market.window_minutes,
            .yes_price = yes_price,
            .no_price = no_price,
            .combined = combined,
            .entry_fees = entry_fees,
            .discount = discount,
            .seconds_remaining = seconds_remaining,
        };

        if (combined > sum_target_) {
            priced.reason = "sum_high";
            priced.score = discount - (combined - sum_target_);
            consider_near_miss(near_miss, priced);
            continue;
        }
        if (discount < min_discount_) {
            priced.reason = "disc_low";
            priced.score = discount;
            consider_near_miss(near_miss, priced);
            continue;
        }

        DumpHedgeSignal signal{
            .market = market,
            .asset = market.asset,
            .yes_token_id = market.yes_token_id,
            .no_token_id = market.no_token_id,
            .yes_price = yes_price,
            .no_price = no_price,
            .combined_price = combined,
            .discount = discount,
            .discount_pct = combined > 0 ? (discount / combined) : 0.0,
            .seconds_remaining = seconds_remaining,
            .timestamp = current_time_ms
        };

        if (!best_signal || signal.discount > best_signal->discount) {
            best_signal = signal;
        }
    }

    const double now_sec = current_time_ms / 1000.0;
    if (!best_signal && near_miss.valid && (now_sec - last_near_miss_log_sec_) >= kDebugLogIntervalSec) {
        last_near_miss_log_sec_ = now_sec;
        std::string detail;
        if (near_miss.reason == "sum_high") {
            detail = fmt::format("sum {:.4f} > target {:.4f} (over {:.4f})",
                                 near_miss.combined, sum_target_, near_miss.combined - sum_target_);
        } else if (near_miss.reason == "disc_low") {
            detail = fmt::format("disc {:.4f} < min {:.4f} (short {:.4f})",
                                 near_miss.discount, min_discount_, min_discount_ - near_miss.discount);
        } else if (near_miss.reason == "min_time") {
            detail = fmt::format("{:.0f}s left < min {:.0f}s", near_miss.seconds_remaining, min_seconds_remaining_);
        } else {
            detail = reject_label(near_miss.reason);
        }

        std::string msg;
        if (near_miss.combined > 0.0) {
            msg = fmt::format(
                "[DH DEBUG] near-miss | {} {}m | YES {:.4f} NO {:.4f} | sum {:.4f} fee {:.4f}/sh disc {:.4f} ({:.2f}%) | {:.0f}s left | {} | {}",
                near_miss.asset, near_miss.window_minutes,
                near_miss.yes_price, near_miss.no_price,
                near_miss.combined, near_miss.entry_fees, near_miss.discount,
                near_miss.discount / near_miss.combined * 100.0,
                near_miss.seconds_remaining, reject_label(near_miss.reason), detail);
        } else {
            msg = fmt::format(
                "[DH DEBUG] near-miss | {} {}m | {:.0f}s left | {} | {}",
                near_miss.asset, near_miss.window_minutes,
                near_miss.seconds_remaining, reject_label(near_miss.reason), detail);
        }
        spdlog::info(msg);
        state_store_.push_telemetry(msg);
    }

    if (best_signal) {
        last_signal_time_[dh_market_key(best_signal->market)] = current_time_ms;
        signals_generated_++;
        std::string msg = fmt::format("DUMP-HEDGE DETECTED [#{}] | {} {}m | YES: {:.3f} NO: {:.3f} | Sum: {:.3f} | Locked: {:.3f}/share",
                                      signals_generated_, best_signal->asset, best_signal->market.window_minutes,
                                      best_signal->yes_price, best_signal->no_price,
                                      best_signal->combined_price, best_signal->discount);
        spdlog::log(spdlog::level::info, msg);
    }

    return best_signal;
}

void DumpHedgeDetector::reset_cooldown(const std::string& asset, double current_time_ms) {
    last_signal_time_[asset] = current_time_ms;
}

} // namespace trading
